### PR TITLE
Add Parental Controls

### LINKED ISSUE #1418: Add a option to disable comments
**Is your feature request related to a usage problem (not a bug)? Please describe.**
I want to spend less time on yt/the internet and that would help

**Describe the solution you'd like to see implemented**
an checkbox to disable comments in the options

**Describe alternatives you've considered**

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by ChunkyProgrammer**
---
Add Parental Controls
---

**Important note**
Please note that only PrestoN is able to merge Pull Requests into master.

**Pull Request Type**
Please select what type of pull request this is:
- [ ] Bugfix
- [x] Feature Implementation

**Related issue**
closes #1418
#1086

**Description**
This PR will allow the following options for parental control:
- [x] Hide comments
- [x] Hide video description
- [x] Hide share button
- [x] Hide unsubscribe button
- [x] Hide live streams
- [x] Hide search bar
- [x] Only show Family Friendly videos/channels

I think seperate PRs would probably be better for the following Parental Control additions (let me know if I should include any in this PR):
- [ ] Password protect settings/app
- [ ] Encrypt local files (settings, profiles, history)
- [ ] Whitelist unblocking (videos, playlists, channels) 
- [ ] Blacklist blocking (videos, playlists, channels, keywords) 

More PRs can also tackle the following:
- [ ] Watchtime control / number of videos to watch per day
- [ ] Always asking for passwords when opening the app
- [ ] Disabling links ( ex: [local api comments](https://github.com/FreeTubeApp/FreeTube/blob/development/src/renderer/components/watch-video-comments/watch-video-comments.js#L253))
- [ ] Disabling certain video categories (ie. Disable all categories but Education & How-To)
- [ ] YouTube kids extractor? (Lot's of work)
- [ ] Hide error messages? (Might not be a good idea)

**Screenshots (if appropriate)**
image
image

**Testing (for code that is not small enough to be easily understandable)**
- Hide Description::
   - Enable Hide Description
   - Navigate to a video
   - Don't see description
 - Hide Live Streams
    - Go to search
    - Search "Lofi"
    - See Lofi Girl
    - Enable "Hide Live Streams"
    - Go to Privacy Settings
    - Click "Clear Search Cache"
    - Search "Lofi"
    - Don't see Lofi Girl
 - Hide Share
   - Navigate to trending page
   - click on 3 dot menu for a video, see share options
   - click on the video, see the share button
    - Enable "Hide Share"
    - Navigate to Trending Page
    - Click 3 dot menu on a video
    - Only see "Marked As Watched"
    - Click on a video
    - See share button is missing
    - Navigate to a playlist
    - See share is missing
 - Hide Comments
    - Go to a video 
    - See "Load Comments" button at the bottom of the video
    - Enable "Hide Comments"
    - Go to a video
    - Don't see "Load Comments" button
  - Hide Search Bar
    - Enable it and search bar disappears
  - Hide Unsubscribe button
    - Navigate to a channel
    - See subscribe button
    - Navigate to a video
    - See subscribe button
    - Navigate to a playlist (i don't remember if playlist shows subscribe button or not)
    - See subscribe button
    - Enable "Hide Unsubscribe button"
    - Navigate to a channel
    - See no Subscribe Button
    - navigate to a video
    - See no Subscribe button
    - Navigate to a playlist
    - See no subscribe button
  - Show Family Friendly Only
    - Search "Fuck"
    - See results
    - Enable "Show Family Friendly Only"
    - Go to Privacy Settings
    - Click Clear Search Cache
    - Search "Fuck"
    - See no results
     This should also work for channels and videos deemed Family Friendly but i have not been able to confirm this (I believe that it may have worked at one point but this PR is so old that I don't remember lol)
 
**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - FreeTube version: 0.13.1

**Additional context**

- There's no meta data for "isFamilyFriendly" on most pages (besides "watch" & "channel") so I added an age restricted component to display when a video or channel isn't family friendly & the toggle is selected (let me know if you can think of a better way of doing this)
image
- Search results can appear empty if only live videos are returned on the first page (maybe some sort of lazy loading of search results can be implemented as an option in the future?)

### CODE REVIEW

**Comment by ChunkyProgrammer** [comment]
@GilgusMaximus I know you mentioned that you were willing to collab with someone on parental controls. I am curious what you think of this PR so far. I feel like the options being introduced can go into distraction free settings but if you think a parental-control-settings component would be better let me know.

**Comment by ChunkyProgrammer** [comment]
Also, I have not been able to test any of the code yet as I am getting an electron build error (unrelated to my changes) so my code might not be as good as it should be
image

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Is #1191 also related?

**Comment by ChunkyProgrammer** [comment]
> Is #1191 also related?

I think it would work better as a seperate PR so this one can be tested, reviewed & merged faster. Atm I plan on finishing this PR and getting it merged then I'll work on the blacklist. The blacklist will probably be very similar to the whitelist so I'll probably take a look at that as well (it shouldn't take too long to implement once I finish the blacklist). Unless a dev thinks otherwise, I'll save that feature for a future PR

**Comment by GilgusMaximus** [comment]
So, sorry for waiting so long for a response, I wasn't able to do anything for a week. Additionally, I am still unable to check any code, so sorry for delaying all the PRs.

I think it is a good idea to split the kids safety into 2 PR's as you suggested, with exactly this split. Because now we can test the UI hiding and then add the additional settings not related to UI.

My suggestion for the setting categories:
- Hide comments: distraction free
- Hide video description: distraction free
- Hide share button: distraction free
- Hide unsubscribe button: parental control, bc you need this as a normal user to manage everything easily
- Hide live streams: distraction free
- Hide search bar: parental controls, bc no normal user would disable the search bar, because then they cannot search anything anymore.
- Only show Family Friendly videos/channels: parental controls, pretty obvious


In regards to the second PR with the other settings, some comments:
- Password protect settings/app
- Encrypt local files (settings, profiles, history) - Here we have to find a way to encrypt and decrypt the data without asking for a password all the time. 
- Whitelist blocking (videos, playlists, channels)
- Blacklist blocking (videos, playlists, channels, keywords) - these also close other PR's, which asked for this. 

So far everything looks good.

Some feature ideas one can add later would be things like watchtime control / number of videos to watch per day, always asking for passwords when opening the app etc...

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Hi @ChunkyProgrammer, i know ur assigned to #95 but I'm just asking to be sure. Is #95 related?

**Comment by ChunkyProgrammer** [comment]
> Hi @ChunkyProgrammer, i know ur assigned to #95 but I'm just asking to be sure. Is #95 related?

Yeah, it'll be included in the next PR

**Comment by peepopoggers** [comment]
Hi, what is the progress with this?

**Comment by ChunkyProgrammer** [comment]
> Hi, what is the progress with this?

Waiting for feedback/review

File: src/renderer/components/ft-list-video/ft-list-video.vue [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <template>
   <div
+    v-if="(!isLive && duration !== '0:00') || (!hideLiveStreams)"
```
Comment (peepopoggers) [inline src/renderer/components/ft-list-video/ft-list-video.vue:3]: I am confused on this logic. Is this checking if the video is a live stream correct?
I may be completely wrong but does the if statement not trigger if: `(!isLive && duration !== '0:00')` is false?

**Review by peepopoggers — COMMENTED** [review]
Looked over everything and looks great. Just check my one question I had about some logic. Other than that LGTM.

File: src/renderer/components/ft-list-video/ft-list-video.vue (hunk 1 above)
Comment (ChunkyProgrammer) [reply src/renderer/components/ft-list-video/ft-list-video.vue:3]: I should probably simplify the logic but it's basically 

```
if not a live stream or live streams aren't hidden then:
     show the video item
else
     display "Live Streams are hidden"

```

IIRC this code only matters for the Most Popular page
image

**Comment by jasonhenriquez** [comment]
I am going to look at this closely. However, be mindful that we should and will not use the approval process for a PR of this size.

File: src/renderer/components/ft-list-video/ft-list-video.vue (hunk 1 above)
Comment (peepopoggers) [reply src/renderer/components/ft-list-video/ft-list-video.vue:3]: My brain just can't get around it xD but if there are no logical issues in testing then that is fine.

**Comment by peepopoggers** [comment]
@jasonhenriquez Agreed, however our reviews will be useful for whoever merges the pr. 
From what I reviewed, all code looked consistent with other similar elements.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@ChunkyProgrammer please resolve merge conflicts in order to get this mergerd.

**Comment by peepopoggers** [comment]
Lets get things moving with this PR. @jasonhenriquez do you think we should ask preston about this?

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@ChunkyProgrammer merge conflicts :(

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
lint failed :O

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I think we can just start reviewing this because we already reviewed PR's what were bigger then this one.

**Comment by peepopoggers** [comment]
Even though this PR is sizable, it is not "risky" because it just implements already existing behavior just over again. I will review this when I have the chance.

**Comment by ChunkyProgrammer** [comment]
> @ChunkyProgrammer i think u forgot to put closes behind 1086 in the template

This pr doesn't password protect settings

File: src/renderer/views/Channel/Channel.vue [hunk 2]
```diff
@@ -7,7 +7,7 @@
       :fullscreen="true"
     />
     <ft-card
-      v-else
+      v-else-if="(isFamilyFriendly==true || !showFamilyFriendlyOnly)"
```
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.vue:10]:
```suggestion
      v-else-if="(isFamilyFriendly || !showFamilyFriendlyOnly)"
```

File: src/renderer/views/Channel/Channel.vue [hunk 3]
```diff
@@ -112,7 +113,7 @@
       </div>
     </ft-card>
     <ft-card
-      v-if="!isLoading"
+      v-if="!isLoading && (isFamilyFriendly==true || !showFamilyFriendlyOnly)"
```
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.vue:116]:
```suggestion
      v-if="!isLoading && (isFamilyFriendly || !showFamilyFriendlyOnly)"
```

**Comment by ChunkyProgrammer** [comment]
> > Edit: i'll do an update to the testing section
> 
> Alright i'll wait for that before testing.

Updated the Testing Section

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM!

File: src/renderer/components/distraction-settings/distraction-settings.js [hunk 4]
```diff
@@ -45,6 +45,18 @@ export default Vue.extend({
     },
     hideActiveSubscriptions: function () {
       return this.$store.getters.getHideActiveSubscriptions
+    },
+    hideDescription: function () {
+      return this.$store.getters.getHideDescription
+    },
+    hideComments: function () {
+      return this.$store.getters.getHideComments
```
Comment (PikachuEXE) [inline src/renderer/components/distraction-settings/distraction-settings.js:53]:
```suggestion
    hideVideoComments: function () {
      return this.$store.getters.getHideVideoComments
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

**Comment by PikachuEXE** [comment]
Why does "Hide Share" mean?
I can see it hides share button, but also copy video URL/open video in browser

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
That is intended. Remember this is in the parental controls. If a child can see the open in browser/copy url then they can also share the video.

**Comment by PikachuEXE** [comment]
"Hide Sharing Features"? "Hide Sharing Actions"?
"Hide Share" read strange to me

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> "Hide Sharing Features"? "Hide Sharing Actions"? "Hide Share" read strange to me

I agree. Hide Sharing Action sounds good!

File: src/renderer/components/distraction-settings/distraction-settings.js (hunk 4 above)
Comment (ChunkyProgrammer) [reply src/renderer/components/distraction-settings/distraction-settings.js:53]: there's a property called hideCommentLikes and I feel like this change will make the code more inconsistent.

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM

**Review by PikachuEXE — DISMISSED** [review]

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@absidue could u review this when u have time :) we need one moreeee

**Review by absidue — CHANGES_REQUESTED** [review]

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]

**Review by PikachuEXE — APPROVED** [review]

**Review by absidue — APPROVED** [review]

**Comment by Joshndroid** [comment]
Is this currently merged or still not quite? Kinda confused. With a fresh download I am unable to see the Parental controls as yet in the settings.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> Is this currently merged or still not quite? Kinda confused. With a fresh download I am unable to see the Parental controls as yet in the settings.

What did u download?

**Comment by Joshndroid** [comment]
> > Is this currently merged or still not quite? Kinda confused. With a fresh download I am unable to see the Parental controls as yet in the settings.
> 
> @Joshndroid What did u download?

I use this as a flatpak on my linux machine and dont see it
I also just did a fresh windows exe download and clicked through each settings menu and dont see it either.

**Comment by absidue** [comment]
@Joshndroid Which version did you download?

**Comment by Joshndroid** [comment]
my flatpak is v0.16.0 Beta
the windows is the same

**Comment by absidue** [comment]
@Joshndroid No wonder you aren't seeing it 0.16.0 was released in February, this was merged in June. You need to download a nightly build of the development branch. There is no flatpack nightly so you will need to use one of the other installation methods on Linux. If you are happy dealing with potentially buggy experimental versions, you can download the latest nightly for the development branch [here](https://github.com/FreeTubeApp/FreeTube/actions/runs/2729648858).

**Comment by Joshndroid** [comment]
perfect che

> @Joshndroid No wonder you aren't seeing it, 0.16.0 was released in February, this was merged in June. You need to download a nightly build of the development branch. There is no flatpack nightly so you will need to use one of the other installation methods on Linux. If you are happy dealing with potentially buggy experimental versions, you can download the latest nightly for the development branch [here](https://github.com/FreeTubeApp/FreeTube/actions/runs/2729648858).

perfect thanks @absidue

**Pull request merged**